boot/bootutil/src/swap_offset.c

```diff
@@ -333,34 +336,35 @@ int boot_slots_compatible(struct boot_loader_state *state)
     size_t sector_sz_pri = 0;
     size_t sector_sz_sec = 0;
     size_t i;
-    size_t num_usable_sectors_pri;
+    size_t num_usable_sectors;
 
     num_sectors_pri = boot_img_num_sectors(state, BOOT_PRIMARY_SLOT);
     num_sectors_sec = boot_img_num_sectors(state, BOOT_SECONDARY_SLOT);
-    num_usable_sectors_pri = app_max_sectors(state);
+    num_usable_sectors = app_max_sectors(state);
 
-    if ((num_sectors_pri != num_sectors_sec) &&
-            ((num_sectors_pri + 1) != num_sectors_sec) &&
-            ((num_usable_sectors_pri + 1) != (num_sectors_sec))) {
+    if (num_sectors_pri != num_sectors_sec &&
+        (num_sectors_pri + 1) != num_sectors_sec &&
+        num_usable_sectors != (num_sectors_sec - 1)) {
```

Comment:
Hm. The allowed primary will be always at least one page shorter than secondary, because that page is used for swapping, so:
```
num_sectors_pri - 1 >= num_sectors_sec
```
is the first condition.
From that we have sectors really available for image as:
```
num_usable_sectors = num_sectors_sec - int((trailer_sz + PAGE_SIZE  - 1) / PAGE_SIZE)
```
That is how much could be uploaded if `num_sectors_pri - 1 >= num_sectors_sec`. Secondary slot sets the limit.

If we have
```
num_sectors_pri - 1 < num_sectors_sec
```
then:
```
num_usable_sectors = (num_sectors_pri - 1) - int((trailer_sz + PAGE_SIZE  - 1) / PAGE_SIZE)
```
That is how much could be swapped, the primary slot sets the limit.